### PR TITLE
gh-124984: Fix `ssl` thread safety

### PR DESCRIPTION
**Pull request by ZeroIntensity**
As it turns out, OpenSSL doesn't like being called in multiple threads. This adds a per-socket (and per-context and per-session) lock for all OpenSSL calls.

* Issue: gh-124984

### CODE REVIEW

File: Lib/test/test_ssl.py
Comment (JelleZijlstra) [inline Lib/test/test_ssl.py:2822]: Missing a `global` statement so this doesn't do anything. Also, let's use a finally block to switch this back.

File: Modules/_ssl.c
Comment (JelleZijlstra) [inline Modules/_ssl.c]: If I'm reading the code right the error state is stored in a thread-local: https://github.com/openssl/openssl/blob/76c4f0e8ea6e885b2b0727c43778fe54ae224135/crypto/err/err.c#L672

File: Lib/test/test_ssl.py
Comment (ZeroIntensity) [reply Lib/test/test_ssl.py:2822]: Ah, right. I'll deal with this in the morning.

File: Modules/_ssl.c
Comment (JelleZijlstra) [inline Modules/_ssl.c]: The PyLong calls could be put outside the lock.

File: Modules/_ssl.c
Comment (JelleZijlstra) [inline Modules/_ssl.c]: Same here

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, that's a lot of locks.

File: Modules/_ssl.c
Comment (picnixz) [inline Modules/_ssl.c]:
```suggestion
#ifdef Py_GIL_DISABLED
#  define PySSL_LOCK(OBJ) 	PyMutex_Lock(&(OBJ)->lock)
#  define PySSL_UNLOCK(OBJ)	PyMutex_Unlock(&(OBJ)->lock)
#else
#  define PySSL_LOCK(OBJ)
#  define PySSL_UNLOCK(OBJ)
#endif
```

File: Modules/_ssl.c
Comment (picnixz) [inline Modules/_ssl.c]:
```suggestion
```

File: Modules/_ssl.c
Comment (picnixz) [inline Modules/_ssl.c]: Technically, it's not even possible to lock it since the lock depends on an attribute that is still not yet initialized.

File: Modules/_ssl.c
Comment (picnixz) [inline Modules/_ssl.c]: The `PyBool_FromLong` could probably be outside of the lock as well.

File: Lib/test/test_ssl.py
Comment (ZeroIntensity) [reply Lib/test/test_ssl.py:2822]: Fixed.

File: Modules/_ssl.c
Comment (ZeroIntensity) [reply Modules/_ssl.c]: Ok, I updated the comment.

File: Modules/_ssl.c
Comment (ZeroIntensity) [reply Modules/_ssl.c]: Eh, I don't think it's worth changing. I don't want somebody to freak out in the future because this function is missing locks.

Comment (picnixz) [reply Modules/_ssl.c]: My point was more on the fact that you cannot even use the macro since `self->lock` is not even set. Having "no need to lock it" may indicate that it's *possible* to use the macro which is not the case.

Comment (ZeroIntensity) [reply Modules/_ssl.c]: I guess I'll move the comment.

**Review by picnixz — DISMISSED** [review]

**Comment by picnixz** [comment]
The failures seem related since they are for the free-threaded build.

**Comment by ZeroIntensity** [comment]
I'll investigate. I unfortunately don't have Windows available to test this on :(

**Comment by picnixz** [comment]
One of the failure is: "connection handling failed: RuntimeError("can't start new thread")"

**Comment by ZeroIntensity** [comment]
Ah, we just have too many threads in the tests.

**Comment by ZeroIntensity** [comment]
I removed some of the threads for functions that don't use a shared `testing_context` as well.

**Comment by ZeroIntensity** [comment]
CI is all good!

File: Lib/test/test_ssl.py
Comment (picnixz) [inline Lib/test/test_ssl.py:2835]: Can you perhaps add a comment saying that one should not just add too many threads? that way we'll know in the future that there is kind of a limit for those tests.

Comment (ZeroIntensity) [reply Lib/test/test_ssl.py:2835]: Added.

**Review by picnixz — APPROVED** [review]

**Comment by ZeroIntensity** [comment]
Hmm, I would create a new issue. I ran the tests from this PR (as well as Jelle's reproducer) on a GIL-ful build, and it seems to work.

**Comment by AA-Turner** [comment]
Thanks @ZeroIntensity, indeed a different issue -- see #125017.

A

File: Misc/NEWS.d/next/Library/2024-10-04-22-43-48.gh-issue-124984.xjMv9b.rst
Comment (gpshead) [inline Misc/NEWS.d/next/Library/2024-10-04-22-43-48.gh-issue-124984.xjMv9b.rst:1]: Mention that this is for the experimental free-threaded build, otherwise people would read this and assume there was a problem with our normal stable build's ssl module.  Also mention how it was fixed (a list of things that now have locks that did not before? or at least summarize that locks around OpenSSL operations were added).

File: Modules/_ssl.c
Comment (gpshead) [inline Modules/_ssl.c]: Perhaps add a note to the ssl module documentation about a threading restriction on do_handshake()?

Comment (ZeroIntensity) [reply Modules/_ssl.c]: I don't think it's needed, nobody should be trying to call `do_handshake()` twice, let alone across multiple threads.

File: Misc/NEWS.d/next/Library/2024-10-04-22-43-48.gh-issue-124984.xjMv9b.rst
Comment (ZeroIntensity) [reply Misc/NEWS.d/next/Library/2024-10-04-22-43-48.gh-issue-124984.xjMv9b.rst:1]: Added.

File: Modules/_ssl.c
Comment (vstinner) [inline Modules/_ssl.c]: Would it make sense to reuse `PyObject.ob_mutex`, rather than adding a second lock per object?

Comment (ZeroIntensity) [reply Modules/_ssl.c]: Hmm, I'm not sure. I was under the impression that weird things happened to `ob_mutex` if running inside a critical section. Though, I couldn't find anything else in the stdlib that uses `ob_mutex` manually, I think we're better off just using our own lock (the point of `PyMutex` is to be tiny, anyway).

**Comment by efimov-mikhail** [comment]
I'm not sure, but this issue may be related to the same topic:
https://github.com/python/cpython/issues/125038

**Comment by ZeroIntensity** [comment]
It's not, that's a bug in the eval loop.

File: Modules/_ssl.c
Comment (colesbury) [inline Modules/_ssl.c]: Don't lock in `tp_dealloc` handlers. Only one thread can deallocate an object.

**Review by colesbury — COMMENTED** [review]
* We generally prefer using critical sections via Argument Clinic `@critical_section` annotations because it's easier to maintain. See https://devguide.python.org/development-tools/clinic/#how-to-use-critical-sections-with-argument-clinic. If you go this route, you'd probably want to convert the get/sets to use Argument Clinic first.
* I think there's going to be a limit to how "safe" we can make this given the places where we already released the GIL and what things we want to lock. For example, the OpenSSL docs say that it's [not safe to modify factory objects](https://docs.openssl.org/master/man7/openssl-threads/#copyright:~:text=it%20is%20not%20safe%20to%20modify%20the%20factory%20object%20after%20it%20has%20been%20used%20to%20create%20other%20objects), like `SSL_CTX` after they are used to create other objects. It seems like `urllib3` is [doing that](https://github.com/urllib3/urllib3/blob/59bbda7e1543d2e8b4b65c3d8075152d664c43ca/src/urllib3/util/ssl_.py#L467).

File: Modules/_ssl.c
Comment (ZeroIntensity) [reply Modules/_ssl.c]: Fixed.

**Comment by ZeroIntensity** [comment]
I would rather just stick with this approach for now. With this, while it's more maintenance, locks are only held for the duration of the OpenSSL calls rather than for the whole call -- `ssl` is something that will likely be run in parallel a lot, so it would be nice to avoid as much contention as possible.

**Comment by colesbury** [comment]
> ssl is something that will likely be run in parallel a lot, so it would be nice to avoid as much contention as possible.

I don't think that matters here. The locks are on individual object instances. Different threads locking different SSL connections won't contend.

**Comment by ZeroIntensity** [comment]
> The locks are on individual object instances.

Yeah, that's what I mean. The locks are per-object here, so we can have methods executing in "more parallel" than with a critical section, because the per-object locks are held for shorter periods of time. I guess I can switch this over to critical sections if it's really needed, but that will make this PR much bigger (it would be nice to backport this to 3.13).

**Review by JelleZijlstra — APPROVED** [review]
Looks good to me, but will leave up to @colesbury to decide which kind of locking to use.

**Comment by ZeroIntensity** [comment]
Thank you, Jelle!

I've marked this as `DO-NOT-MERGE` for the time being. Apparently, some OpenSSL calls can execute arbitrary Python code via callbacks, which will result in deadlocks. I'll have to switch at least some of the methods over to critical sections, if not all.

**Comment by vstinner** [comment]
Moving towards critical section sounds like the right move. It sounds less likely to cause deadlock issue.

**Comment by ZeroIntensity** [comment]
Build failures seem to be caused by a docstring macro for `owner` and `session` getting redefined. Possibly a bug in clinic?

File: Lib/test/test_ssl.py
Comment (vstinner) [inline Lib/test/test_ssl.py]: Would you mind to elaborate? Explain shortly the purpose of the test?

File: Lib/test/test_ssl.py
Comment (vstinner) [inline Lib/test/test_ssl.py]: I'm not sure that calling start()+join() reports properly errors to this thread/function.

You should use something like `with catch_threading_exception() as cm:` of `test.support.threading_helper` and check `cm.exc_value`.

File: Lib/test/test_ssl.py
Comment (vstinner) [inline Lib/test/test_ssl.py:2837]: Maybe store func in the function, to be able to pass it to subTest() below? Something like:

```py
thread = Thread(target=func)
thread.func = func
threads.append(thread)
```

and then:

```
with self.subTest(thread=thread, func=thread.func):
```

Comment (ZeroIntensity) [reply Lib/test/test_ssl.py:2837]: Hmm, why? It would just add more complexity.

Comment (vstinner) [reply Lib/test/test_ssl.py:2837]: To give more function if an exception is raised.

But I'm not sure why you use subTest() to call start() and join(), these functions should not fail.

Comment (ZeroIntensity) [reply Lib/test/test_ssl.py:2837]: Theoretically, they could raise a `MemoryError` or something, but I guess the subtest isn't useful. I'll remove it.

**Comment by vstinner** [comment]
@ZeroIntensity:
> Apparently, some OpenSSL calls can execute arbitrary Python code via callbacks, which will result in deadlocks. I'll have to switch at least some of the methods over to critical sections, if not all.

How does it work with callback? For example, `SSLContext.load_cert_chain(password=function)` calls `function()` in the criticial section. What if the callback calls another SSLContext method? Does it hang?

I tried the following code and it doesn't hang but I don't understand why :-)

```py
import _ssl
import os.path
ctx = _ssl._SSLContext(_ssl.PROTOCOL_TLS_CLIENT)

KEY_PASSWORD = "somepass"
certfile = os.path.join("Lib", "test", "certdata", "keycert.passwd.pem")
def password_func():
    print("*callback*")
    check_hostname = ctx.check_hostname
    print(f"{check_hostname=}")
    return KEY_PASSWORD
ctx.load_cert_chain(certfile, password=password_func)
```

**Comment by ZeroIntensity** [comment]
Critical sections are a special locking mechanism that prevents lock ordering deadlocks like this. PEP 703 probably has more information about it.

I *think* this is how it works: when a critical section is activated, it locks the object and appends it to a stack. When a `Py_BEGIN_ALLOW_THREADS` block is reached, all critical sections are deactivated and locks are released. The top-most critical section is re-locked upon reacquiring the thread state (`Py_END_ALLOW_THREADS`), but the prior critical sections are left released.

I'm pretty sure it was @colesbury who designed it

**Comment by vstinner** [comment]
I found documentation at: https://docs.python.org/dev/c-api/init.html#python-critical-section-api

> The critical section API provides a **deadlock avoidance** layer on top of per-object locks for free-threaded CPython.

It seems that it just works :-)

**Review by vstinner — APPROVED** [review]
LGTM.

This change basically adds `@critical_section` on all `ssl` methods. The change looks exhaustive. In the worst case, if a method was missed, we can protect it later with a critical section.

Python callbacks don't deadlock thanks for the dark magic of the critical section API.

The change also adds a stress-test on critical sections by running multiple `test_ssl` tests in parallel.

**Comment by ZeroIntensity** [comment]
I want to run this with build bots before merging, so there aren't any surprises.

**Review by corona10 — APPROVED** [review]
lgtm (as the member who wrote `@getter`, `@setter` and related `@critical_section` logics)

**Comment by vstinner** [comment]
> buildbot/AMD64 Arch Linux TraceRefs PR — Build done.
> buildbot/x86 Debian Installed with X PR — Build done.
> buildbot/x86 Debian Non-Debug with X PR — Build done. 

These 3 failures are known and unrelated.

**Comment by ZeroIntensity** [comment]
> These 3 failures are known and unrelated.

I figured. Let's get this merged :)

**Comment by vstinner** [comment]
>  Tests / Thread sanitizer (free-threading)

test_ssl fails with:

```
0:03:08 load avg: 6.09 [15/22/1] test_ssl failed (1 failure) (33.9 sec) -- running (1): test_socket (1 min 15 sec)
test test_ssl failed -- Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/test/test_ssl.py", line 2848, in test_ssl_in_multiple_threads
    self.assertIsNone(cm.exc_value)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
AssertionError: SkipTest("resource 'walltime' is not enabled") is not None
```

**Comment by ZeroIntensity** [comment]
Oh, I didn't see that one, nevermind. I'll fix that in the next hour or so.

**Comment by vstinner** [comment]
Merged, thank you! I replaced "Fix" with "Enhance" in the commit message: "Enhance ssl thread safety" :-)

**Comment by vstinner** [comment]
Remaining question: should we backport this *enhancement* (bugfix?) to Python 3.13?

**Comment by ZeroIntensity** [comment]
> Remaining question: should we backport this _enhancement_ (bugfix?) to Python 3.13?

Yeah, I had the same thought. I'll leave that decision to @Yhg1s

**Comment by vstinner** [comment]
Without this change, the `ssl` module is not usable on a Free Threaded build (it does easily crash). The change only affects the Free Threaded build: adding `@critical_section` does nothing in the regular build, and the added test is ignored on regular Python (only run on Free Threading). I would be fine with no backporting it, since Free Threading remains experimental.

**Comment by colesbury** [comment]
I think that fixing the SSL module crash in the 3.13 free threading build is important -- lots of basic tasks around HTTP requests are likely to crash without it.

I don't think "lines of code changed" is not a good measure of the complexity here -- the `Modules/_ssl.c` changes are mostly mechanical, and the only behavioral changes is additional locking in the free threading build. 

If backporting this PR is a nonstarter than we should consider a smaller, more targeted change that only adds `@critical_section` to `SSLContext` methods (without changing getters and setters). That seems less ideal to me, but should fix the common SSL crash.

**Comment by colesbury** [comment]
Also, thank you @ZeroIntensity for fixing this bug and @vstinner and @corona10 for review it.

**Comment by vstinner** [comment]
@ZeroIntensity: Automated backport failed. Would you mind to backport the change manually?

With a backport, it might be easier to take a decision on fixing 3.13 or not.

**Comment by ZeroIntensity** [comment]
Yeah, I can do it later today.

**Comment by vstinner** [comment]
@Yhg1s: Are you still against the backport to 3.13 after @colesbury's comment?

**Comment by Yhg1s** [comment]
I'm okay with a backport of just the @critical_section changes (since those expand to nothing in the normal build, and the free-threaded build is experimental anyway). It's the larger refactorings that worry me.

**Comment by vstinner** [comment]
> It's the larger refactorings that worry me.

Other changes are tests and changes to use the code declared with `@critical_section`.

**Comment by ZeroIntensity** [comment]
There isn't any other refactoring going on here, I just had to switch the getters and setters over to AC for the critical section. Another issue is that not backporting this to 3.13 will also hurt any automatic backports for `ssl` in the future.

File: Modules/clinic/_ssl.c.h
Comment (youknowone) [inline Modules/clinic/_ssl.c.h:1154]: The symbol `_ssl__SSLContext_security_level_set` is used here, but nowhere added

Comment (ZeroIntensity) [reply Modules/clinic/_ssl.c.h:1154]: This is generated clinic code. I'm pretty sure this route is never taken, and the macro right after this is used instead.

Comment (youknowone) [reply Modules/clinic/_ssl.c.h:1154]: Oh, thank you. I didn't know that

**Pull request merged**
